### PR TITLE
MDC data updates 

### PR DESCRIPTION
**Pull request by EthanMarx**
Exact clone of #352. Plan is to merge into mdc-dev branch so that we can start adding fixes.

### CODE REVIEW

**Comment by wbenoit26** [comment]
Looks like an `import` statement in `test_export` needs to be updated

File: projects/sandbox/datagen/datagen/scripts/timeslide_waveforms.py [hunk 1]
```diff
@@ -238,17 +249,18 @@ def deploy(
     # to accumulate desired background livetime
     state_flags = [f"{ifo}:{state_flag}" for ifo in ifos]
     segments = query_segments(state_flags, start, stop, min_segment_length)
-    shifts_required = utils.calc_shifts_required(segments, Tb, max(shifts))
+    shifts_required = utils.get_num_shifts(segments, Tb, shifts[1])
```
Comment (wbenoit26) [inline projects/sandbox/datagen/datagen/scripts/timeslide_waveforms.py:252]: Was there a reason to switch from `max(shifts)`?

Comment (EthanMarx) [reply projects/sandbox/datagen/datagen/scripts/timeslide_waveforms.py:252]: No good call that got lost in the shuffle

File: projects/sandbox/train/train/train.py [hunk 2]
```diff
@@ -212,43 +209,25 @@ def main(
...
-        rescaler.fit(*background)
-        valid_waveforms = train_utils.threshold_snrs(
-            valid_waveforms,
-            snr_thresh,
-            sample_rate,
-            rescaler.background,
-            highpass,
-        )
+    if valid_waveforms is not None:
+        valid_fname = background_fnames.pop(-1)
```
Comment (wbenoit26) [inline projects/sandbox/train/train/train.py:228]: Did we switch to having the validation dataset just be the last segment from the training time?

File: projects/sandbox/train/train/train.py [hunk 3]
```diff
@@ -266,10 +245,14 @@ def main(
...
-            sample_rate,
+            psd_estimator=psd_estimator,
+            whitener=whitener,
+            snr_thresh=snr_thresh,
+            highpass=highpass,
+            sample_rate=sample_rate,
             stride=valid_stride,
             injection_stride=4,
-            kernel_length=kernel_length,
+            kernel_length=sample_length,
```
Comment (wbenoit26) [inline projects/sandbox/train/train/train.py:255]: Worth thinking about how to get our naming conventions clear and consistent here, though certainly not top priority.

File: projects/sandbox/train/train/validation.py [hunk 4]
```diff
@@ -191,32 +196,69 @@ def postprocess(self, preds: torch.Tensor) -> torch.Tensor:
...
+
+        snrs = gw.compute_network_snr(waveforms, psds, self.sample_rate, mask)
+        target_snrs = snrs.clamp(self.snr_thresh, 1000)
+        weights = target_snrs / snrs
+        return waveforms * weights.view(-1, 1, 1)
+
+    def inject(self, X: torch.Tensor, psds: torch.Tensor) -> torch.Tensor:
+        # downsample the background batch to give us
+        # the desired stride between injections
         X = X[:: self._injection_step]
```
Comment (wbenoit26) [inline projects/sandbox/train/train/validation.py:216]: I don't understand this step. X has the shape `(batch_size, num_ifos, background_length)`, right? So we're taking every `self._injection_step` kernel?

File: projects/sandbox/train/train/validation.py [hunk 5]
```diff
@@ -191,32 +196,69 @@ def postprocess(self, preds: torch.Tensor) -> torch.Tensor:
...
+        X = X[: len(waveforms)]
+        self._injection_idx += len(waveforms)
+
+        # threshold the SNRs of the injections to the desired value.
+        waveforms = self.threshold_snrs(waveforms, psds)
+
+        # now cut out a window symmetrically about the
+        # coalescence time and inject it into the background
+        kernel_size = X.shape[-1]
+        start = waveforms.shape[-1] // 2 - kernel_size // 2
```
Comment (wbenoit26) [inline projects/sandbox/train/train/validation.py:236]: We're always putting the coalescence point in the middle now? Will that be a temporary change?

File: projects/sandbox/infer/infer/deploy.py [hunk 6]
```diff
@@ -154,9 +157,11 @@ def main(
 
     # spin up triton server
     logging.info("Launching triton server")
-    with serve(model_repo_dir, image, wait=True):
+    server_log = log_dir / "server.log"
+    with serve(model_repo_dir, image, log_file=server_log, wait=True):
         # launch inference jobs via condor
         logging.info("Server online")
+        time.sleep(1)
```
Comment (wbenoit26) [inline projects/sandbox/infer/infer/deploy.py:164]: What bug does this solve?

File: projects/sandbox/infer/infer/callback.py [hunk 7]
```diff
@@ -116,8 +119,9 @@ def cluster(self, y) -> TimeSlideEventSet:
         # subtract off the time required for
         # the coalescence to exit the filter
         # padding and enter the input kernel
-        # to the neural network
-        t0 = self.start - self.fduration / 2
+        # to the neural network, and account
+        # for the PSD burn-in that has been removed
+        t0 = self.start - self.fduration / 2 + self.psd_length
```
Comment (wbenoit26) [inline projects/sandbox/infer/infer/callback.py:124]: Should `psd_length` be subtracted here? Seems like it should shift things in the same direction as `fduration`, though I'm having a hard time thinking this through.

**Review by wbenoit26 — COMMENTED** [review]
I don't think I totally get what's going on in `snapshotter.py`, but I've looked over everything else.

File: projects/sandbox/datagen/datagen/scripts/timeslide_waveforms.py [hunk 8]
```diff
@@ -289,10 +301,10 @@ def deploy(
     logging.info(f"Merging output waveforms to file {waveform_fname}")
     LigoResponseSet.aggregate(waveform_files, waveform_fname, clean=True)
 
-    params_fname = writedir / "rejected-parameters.h5"
-    param_files = list(outdir.rglob("rejected-parameters.h5"))
-    logging.info(f"Merging rejected parameters to file {params_fname}")
-    InjectionParameterSet.aggregate(param_files, params_fname, clean=True)
+    # params_fname = writedir / "rejected-parameters.h5"
```
Comment (alecgunny) [inline projects/sandbox/datagen/datagen/scripts/timeslide_waveforms.py:304]: Some of this commented stuff that we don't need because we're not rejecting params anymore, just go ahead and delete

File: projects/sandbox/pyproject.toml [hunk 9]
```diff
@@ -1,12 +1,14 @@
 [tool.pinto]
 steps = [
-    "datagen:generate-background",
-    "datagen:generate-waveforms",
-    "datagen:generate-glitches",
-    "datagen:deploy-timeslide-waveforms",
+    #"datagen:generate-background",
```
Comment (alecgunny) [inline projects/sandbox/pyproject.toml:3]: undo comments

File: projects/sandbox/pyproject.toml [hunk 10]
```diff
@@ -90,46 +96,50 @@ minimum_frequency = "${base.minimum_frequency}"
...
-batch_size = 512
+
+batch_size = 640
+
+
 max_epochs = 200
-max_lr = 1.33e-4
-lr_ramp_epochs = 9
+max_lr = 1e-3
+lr_ramp_epochs = 60
```
Comment (alecgunny) [inline projects/sandbox/pyproject.toml:122]: Just confirming whether you want this or not

File: projects/sandbox/train/train/train.py (hunk 2 above)
Comment (alecgunny) [reply projects/sandbox/train/train/train.py:228]: Yeah exactly, but we do timeslides so it doesn't affect the amount of background we generate

File: projects/sandbox/train/train/validation.py (hunk 4 above)
Comment (alecgunny) [reply projects/sandbox/train/train/validation.py:216]: Yeah the idea here is to put some space in between injections (in time) to avoid spurious correlations

**Review by alecgunny — COMMENTED** [review]
Couple notes but overall think this is worth getting in ASAP and then we can fine tune from there

File: projects/sandbox/export/export/snapshotter.py [hunk 11]
```diff
@@ -96,6 +96,7 @@ def add_streaming_input_preprocessor(
 
     stride = int(sample_rate / inference_sampling_rate)
     state_shape = (1, num_ifos, snapshotter.state_size)
+    print(state_shape)
```
Comment (wbenoit26) [inline projects/sandbox/export/export/snapshotter.py:99]: Remove `print` statement

**Comment by EthanMarx** [comment]
Since we're merging this into `mdc-dev` going to suggest we go ahead despite broken tests. 
Both code and tests will have to change significantly with new `ml4gw` tools like `LocalWhitener` / `Dataloader`, so I dont think its worth reworking them now just to have to again shortly

**Comment by EthanMarx** [comment]
@alecgunny @wbenoit26 Expecting these tests to pass and this should be ready for review.

Recall that this PR is a branch off of #352, which implements chunked data loading and local whitening.

On top of that PR, this branch 
1. Removes the first `psd_length` seconds of data from each segment due to psd burn in. This is also accounted for in the timeslide waveforms generation.
2. Fixes a bug in the inference project where some shifts were failing if there was leftover data smaller than the `batch_size`
3. Standardizes nomenclature -  `kernel_length` refers to length of data in seconds seen by neural network,  `psd_length` refers to length of data in seconds used to create psd for whitening. `kernel_length + fduration` seconds of data are whitened before `fduration / 2` seconds are cropped from each end. 


Important things that to note:
1. We need to write new tests for the train project. It didn't make sense to write these now, since we will soon start using the `ml4gw` dataloader / whitener.
2.  I rewrote the export tests for the new inference deployment but would love a second eye on them
4.  I removed the "hacky" way I was implementing the MDC's version of the SD calculation. I think this should be discussed / implemented in @wbenoit26 PR #362 that is looking to refactor how we do priors.

**Review by alecgunny — APPROVED** [review]
Export tests look good to me. Agree that everything else can (and probably needs to) be done in future add-on PRs
